File: medium/513/find_bottom_left_tree_value.py

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findBottomLeftValue(self, root: TreeNode) -> int:
        
        acc = []
        acc.append((0, root))
        selected = (0, root.val)
        
        while len(acc) > 0:
            row, node = acc.pop()
            
            if node.right is not None:
                acc.append((row + 1, node.right))

            if node.left is not None:
                acc.append((row + 1, node.left))
            else:
                if row > selected[0]:
                    selected = (row, node.val)
        
        return selected[1]

File: medium/513/find_bottom_left_tree_value.cpp

#include <utility>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findBottomLeftValue(TreeNode* root) {
        
        vector<std::pair<int,TreeNode*>> acc;
        std::pair<int,int> selected;
        
        acc.push_back(std::make_pair(0, root));
        selected = std::make_pair(0, root->val);
        
        while (acc.size() > 0) {
            std::pair<int,TreeNode*> p = acc.back();
            acc.pop_back();
            
            int row = p.first;
            TreeNode * node = p.second;
            TreeNode* left = node->left;
            TreeNode* right = node->right;

            if (right != NULL) {
                acc.push_back(std::make_pair(row + 1, right));
            }
            
            if (left != NULL) {
                acc.push_back(std::make_pair(row + 1, left));
            }
            else {
                if (row > selected.first) {
                    selected = std::make_pair(row, node->val);
                }
            }
        }       
            
        return selected.second;
    }
};
